### PR TITLE
Enable compile-time multiplicity enforcement for computed links

### PR DESCRIPTION
**Pull request by elprans**
A computed link should be indistinguishable from a regular link in queries.
This means that just like regular links, a computed link
_must_ resolve to a distinct set.  After this, it is a compile-time error
to use any expression that could not be proved to have multiplicity of at
most 1.  Queries where static inference is impossible should use
`assert_distinct` or `DISTINCT` depending on whether duplicates are
expected.

Fixes: #117

### CODE REVIEW

File: tests/test_edgeql_update.py [hunk 1]
```diff
@@ -2937,10 +2939,10 @@ async def test_edgeql_update_add_dupes_03(self):
                     UPDATE UpdateTestSubSubType
                     FILTER .name = 'add-dupes'
                     SET {
-                        tags += {
+                        tags += assert_distinct({
                             (UPDATE Tag FILTER .name = 'fun' SET {flag := 1}),
                             (UPDATE Tag FILTER .name = 'wow' SET {flag := 2}),
-                        }
+                        })
```
Comment (vpetrovykh) [inline tests/test_edgeql_update.py:2945]: Do we actually *need* an `assert_distinct` here? I thought that `+=` and `-=` already will make linked sets distinct in the end. I mean if I already have `Tag` "wow" linked as `tags`, doing `tags += (select Tag filter .name = 'wow')` should be legal, so who cares if the set after `+=` or `-=` is not distinct if the result must be reconciled anyways?

**Review by vpetrovykh — APPROVED** [review]
LGTM. I do have a question about one of the tests, but it probably doesn't impact the PR overall.

File: edb/edgeql/qltypes.py [hunk 2]
```diff
@@ -172,10 +174,35 @@ def _missing_(cls, name):
...
+        return self is MultiplicityValue.ONE
+
+    def is_many(self) -> bool:
+        return self is MultiplicityValue.MANY
+
+    def is_zero(self) -> bool:
+        return self is MultiplicityValue.ZERO
+
+
+@dataclasses.dataclass(frozen=True, eq=False)
```
Comment (msullivan) [inline edb/edgeql/qltypes.py:193]: Probably worth adding a quick comment here

File: edb/edgeql/compiler/inference/cardinality.py [hunk 3]
```diff
@@ -660,6 +660,16 @@ def __infer_func_call(
         _, arg_upper = arg_card
         return _bounds_to_card(CardinalityBound.ONE, max(arg_upper))
 
+    elif ir.func_shortname == sn.QualName('std', 'assert_distinct'):
```
Comment (msullivan) [inline edb/edgeql/compiler/inference/cardinality.py:663]: Why does this need special handling?

File: edb/pgsql/compiler/relgen.py [hunk 4]
```diff
@@ -2003,6 +2008,143 @@ def process_set_as_existence_assertion(
...
+    #        (CASE WHEN
+    #            <target_set>
+    #            IS DISTINCT FROM
+    #            lag(<target_set>) OVER (ORDER BY <target_set>)
+    #        THEN <target_set>
+    #        ELSE edgedb.raise(ConstraintViolationError))
+    #    FROM
+    #        (SELECT <target_set>, row_number() OVER () AS i) AS q
+    #    ORDER BY
+    #        q.i
```
Comment (msullivan) [inline edb/pgsql/compiler/relgen.py:2049]: ordering by the check_expr too, right?

**Review by msullivan — APPROVED** [review]
This looks pretty solid. I think that fixing the xfails should be pretty simple

File: edb/pgsql/compiler/relgen.py [hunk 5]
```diff
@@ -2003,6 +2012,151 @@ def process_set_as_existence_assertion(
...
+            ),
+        ])
+
+        pathctx.put_path_id_map(newctx.rel, ir_set.path_id, ir_arg_set.path_id)
+
+    aspects = ('value', 'source')
+
+    func_rvar = relctx.new_rel_rvar(ir_set, newctx.rel, ctx=ctx)
+    relctx.include_rvar(stmt, func_rvar, ir_set.path_id,
+                        pull_namespace=False, aspects=aspects, ctx=ctx)
```
Comment (msullivan) [inline edb/pgsql/compiler/relgen.py:2155]: Dropping the `pull_namespace=False` fixes the 4 ISE xfails

**Comment by msullivan** [comment]
Oh, and you need to bump `EDGEDB_CATALOG_VERSION`

**Comment by msullivan** [comment]
Aha, and the problem with linkprops is that `compile_call_arg` always wraps call arguments in a `SelectStmt` for fencing and implicit limit insertion purposes. Compiling operator arguments works a little different, so it dodges that.

File: tests/test_edgeql_update.py (hunk 1 above)
Comment (elprans) [reply tests/test_edgeql_update.py:3178]: This is a good point.  I think we can disable multiplicity checks for `-=` and `+=`

File: edb/edgeql/compiler/inference/cardinality.py (hunk 3 above)
Comment (elprans) [reply edb/edgeql/compiler/inference/cardinality.py:663]: `std::assert_distinct` preserves the cardinality of its argument, it's completely transparent.

File: edb/pgsql/compiler/relgen.py (hunk 4 above)
Comment (elprans) [reply edb/pgsql/compiler/relgen.py:2049]: Right.

File: edb/edgeql/qltypes.py (hunk 2 above)
Comment (elprans) [reply edb/edgeql/qltypes.py:193]: Actually, I think I'll make the structured multiplicity private to inference.  I originally made it this way to minimize the `multiplicity.py` diff, but it had grown very large anyway.

File: tests/test_edgeql_update.py (hunk 1 above)
Comment (1st1) [reply tests/test_edgeql_update.py]: +1

File: edb/pgsql/compiler/relgen.py (hunk 5 above)
Comment (elprans) [reply edb/pgsql/compiler/relgen.py]: Oh, good.

File: edb/edgeql/compiler/inference/utils.py [hunk 6]
```diff
@@ -0,0 +1,63 @@
+#
+# This source file is part of the EdgeDB open source project.
+#
+# Copyright 2008-present MagicStack Inc. and the EdgeDB authors.
```
Comment (1st1) [inline edb/edgeql/compiler/inference/utils.py:4]: Definitely not 2008 ;)

File: edb/pgsql/compiler/relgen.py [hunk 7]
```diff
@@ -2003,6 +2008,143 @@ def process_set_as_existence_assertion(
...
+    # Generate a distinct set assertion as the following SQL:
+    #
+    #    SELECT
+    #        <target_set>,
+    #        (CASE WHEN
+    #            <target_set>
+    #            IS DISTINCT FROM
+    #            lag(<target_set>) OVER (ORDER BY <target_set>)
+    #        THEN <target_set>
+    #        ELSE edgedb.raise(ConstraintViolationError)) AS check_expr
```
Comment (1st1) [inline edb/pgsql/compiler/relgen.py:2045]: Should we add a new class of errors -- RuntimeAssertionError?

File: edb/edgeql/compiler/inference/multiplicity.py [hunk 8]
```diff
@@ -189,21 +193,33 @@ def _infer_shape(
...
+                and irtyputils.is_object(ptrref.out_target)
+            ):
+                ctx.env.schema, ptrcls = irtyputils.ptrcls_from_ptrref(
+                    ptrref, schema=ctx.env.schema)
+                assert isinstance(ptrcls, s_pointers.Pointer)
+                desc = ptrcls.get_verbosename(ctx.env.schema)
+                if not is_mutation:
+                    desc = f"computed {desc}"
                 raise errors.QueryError(
                     f'possibly not a strict set returned by an '
```
Comment (1st1) [inline edb/edgeql/compiler/inference/multiplicity.py:216]: strict -> distinct?

File: edb/pgsql/compiler/relgen.py (hunk 7 above)
Comment (elprans) [reply edb/pgsql/compiler/relgen.py:2045]: Why would that be better than a `ConstraintViolationError` or a `CardinalityViolationError`?

File: edb/edgeql/compiler/inference/multiplicity.py (hunk 8 above)
Comment (elprans) [reply edb/edgeql/compiler/inference/multiplicity.py:216]: Sure.

File: edb/edgeql/compiler/inference/utils.py (hunk 6 above)
Comment (elprans) [reply edb/edgeql/compiler/inference/utils.py:4]: Fixed.

**Pull request merged**